Source/Data/Remotes/ParleyRemote.swift

```diff
@@ -4,9 +4,60 @@ import ObjectMapper
 import UIKit
 import Foundation
 
+
+public typealias ParleyRemoteInterceptor = ParleyRemoteRequestInterceptorProtocol & ParleyRemoteResponseInterceptorProtocol
+
+public protocol ParleyRemoteRequestInterceptorProtocol: AnyObject {
+    func httpRequestWillStart(with request: inout URLRequest) throws
+}
+
+public protocol ParleyRemoteResponseInterceptorProtocol: AnyObject {
+    func httpResponseReceived(response: HTTPURLResponse) throws
+}
+
+internal class ParleyRemoteInterceptorProxy: RequestInterceptor {
+
+    private let interceptors: [ParleyRemoteInterceptor]
+    
+    init(interceptors: [ParleyRemoteInterceptor]) {
+        self.interceptors = interceptors
+    }
+    
+    func adapt(_ urlRequest: URLRequest, for session: Session, completion: @escaping (Result<URLRequest, Error>) -> Void) {
+        guard interceptors.count > 0 else {
+            completion(.success(urlRequest))
+            return
+        }
+        var editableUrlRequest = urlRequest
+        
+        interceptors.forEach {
+            do {
+                try $0.httpRequestWillStart(with: &editableUrlRequest)
+            } catch {
+                completion(.failure(error))
```

Comment:
Is the `throws` needed? In what case would adjusting the request throw. And why is it needed to be able to cancel the request in this case? Cancelling requests may cause side effects for the actual functioning of Parley. So we are looking for whether this is needed or not.

Reply:
In case an interceptor finds out the request has been tampered we would like to to fail. Would a retry mechanism a solution to mitigate the side effects?